platform/SharedPtr.h

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    /**
+     * @brief Assignment operator.
+     * @details Cleanup previous reference and assign new pointer and counter.
+     * @param source Object being assigned from.
+     * @return Object being assigned.
+     */
+    SharedPtr operator=(const SharedPtr& source) {
+        if (this != &source) {
+            // clean up by decrementing counter
+            decrement_counter();
+
+            // assign new values
+            _ptr = source.get();
+            _counter = source.get_counter();
+
+            // increment new counter
+            if (_ptr != NULL) {
+                core_util_critical_section_enter();
+                (*_counter)++;
+                core_util_critical_section_exit();
+            }
+        }
+
+        return *this;
+    }
+
+    /**
+     * @brief Replaces the managed pointer with a new unmanaged pointer.
+     * @param[in] ptr the new raw pointer to manage. 
+     */
+    void reset(T* ptr) {
+        // clean up by decrementing counter
+        decrement_counter();
+
+        if(ptr != NULL) {
+            // allocate counter on the heap so it can be shared
+            _counter = (uint32_t*) malloc(sizeof(uint32_t));
+            *_counter = 1;
+        }
+    }
+
+    /**
+     * @brief Replace the managed pointer with a NULL pointer.
+     */ 
+    void reset() {
+        reset(NULL);
+    }
+
+    /**
+     * @brief Raw pointer accessor.
+     * @details Get raw pointer to object pointed to.
+     * @return Pointer.
+     */
+    T* get() const {
+        return _ptr;
+    }
+
+    /**
+     * @brief Reference count accessor.
+     * @return Reference count.
+     */
+    uint32_t use_count() const {
+        if (_ptr != NULL) {
+            core_util_critical_section_enter();
+            return *_counter;
+            core_util_critical_section_exit();
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * @brief Dereference object operator.
+     * @details Override to return the object pointed to.
+     */
+    T& operator*() const {
+        return *_ptr;
+    }
+
+    /**
+     * @brief Dereference object member operator.
+     * @details Override to return return member in object pointed to.
+     */
+    T* operator->() const {
+        return _ptr;
+    }
+
+    /**
+     * @brief Boolean conversion operator.
+     * @return Whether or not the pointer is NULL.
+     */
+    operator bool() const {
+        return (_ptr != NULL);
+    }
+
+private:
+    /**
+     * @brief Get pointer to reference counter.
+     * @return Pointer to reference counter.
+     */
+    uint32_t* get_counter() const {
+        return _counter;
+    }
+
+    /**
+     * @brief Decrement reference counter.
+     * @details If count reaches zero, free counter and delete object pointed to.
+     */
+    void decrement_counter() {
+        if (_counter) {
+            core_util_critical_section_enter();
+            if (*_counter == 1) {
```

Comment:
Tiny tiny tiny micro-optimization, so feel free to ignore. But comparing with zero is slightly better in terms of code size. Once you hit zero there's also no race conditions possible.